src/js/actions/layoutActions.ts

```diff
@@ -0,0 +1,67 @@
+import {createStandardAction} from 'typesafe-actions';
+import {LayoutRecord, PlaceHolderRecord, placeHolder} from '../store/factory/Layout';
+import {assignId} from '../util/counter';
+import {State} from '../store';
+import {Dispatch} from 'redux';
+import {MarkRecord} from '../store/factory/Mark';
+import {setSignal} from './signalActions';
+import {batchGroupBy} from '../reducers/historyOptions';
+
+export function addLayout (payload: LayoutRecord) {
+  return function(dispatch: Dispatch, getState: () => State) {
+    const id = payload._id || assignId(dispatch, getState());
+    dispatch(baseAddLayout(payload.merge({_id: id}), id));
+  };
+}
+export const baseAddLayout = createStandardAction('ADD_LAYOUT')<LayoutRecord, number>();
+
+
+export function addGrouptoLayout (payload: {group: MarkRecord, dir: string}, id: number) {
```

Comment:
what if we use the more specific type annotation `GroupRecord` instead of `MarkRecord`?